guides/adr/0012-formalize-accredited-provider-relationships.md

```diff
@@ -0,0 +1,380 @@
+          provider_code: accredited_provider.provider_code,
+          description: accrediting_provider_enrichment.Description || ''
+        }
+      end
+    end || []
+  end
+
+  def add_enrichment_errors
+    accrediting_provider_enrichments&.each do |item|
+      provider_code = item.UcasProviderCode
+
+      accredited_provider = recruitment_cycle.providers.find_by(provider_code:)
+
+      if accredited_provider.present? && item.invalid?
+        message = "^Reduce the word count for #{accredited_provider.provider_name}"
+        errors.add :accredited_bodies, message
+      end
+    end
+  end
+
+  def accredited_provider_codes
+    accrediting_provider_enrichments&.map(&:UcasProviderCode) || []
+  end
+
+    ...
+
+class Organisation < ApplicationRecord
+  has_many :organisation_users
+
+  has_many :users, through: :organisation_users
+
+  has_and_belongs_to_many :providers
+
+    ...
+
+class Course < ApplicationRecord
+
+  def accrediting_provider_description
+    return if accrediting_provider.blank?
+    return if provider.accrediting_provider_enrichments.blank?
+
+    accrediting_provider_enrichment = provider.accrediting_provider_enrichments
+                                              .find do |provider|
+      provider.UcasProviderCode == accrediting_provider.provider_code
+    end
+
+    accrediting_provider_enrichment.Description if accrediting_provider_enrichment.present?
+  end
+```
+#### Pros
+
+- No effort expended and other work can be prioritised
+
+#### Cons
+
+- Eases the burden of onboarding developers
+- Leaves the data in an under constrained state
+- Inefficient methods for querying entity relationships
+
+### 2. Create typical self-join has_many through pivot table relationship
+
+```ruby
+class Provider < ApplicationRecord
+  has_many :accredited_accreditations,
+           class_name: 'ProviderAccreditation',
+           foreign_key: :training_provider_id
+
+  has_many :training_accreditations,
+           class_name: 'ProviderAccreditation',
+           foreign_key: :accredited_provider_id
+
+  has_many :accredited_providers,
+           through: :accredited_accreditations,
+           source: :accredited_provider,
+           class_name: 'Provider'
+
+  has_many :training_providers,
+           through: :training_accreditations,
+           source: :training_provider,
+           class_name: 'Provider'
+
+class ProviderAccreditation < ApplicationRecord
+
+```
+
+#### Pros
+
+- Query the training provider / accredited provider relationships through typical relational associations
+- Index the columns used to make these queries.
+- More formally distinguish accredited and accrediting providers.
+- Database validations (if desired) on which accredited providers can accredit which courses.
+- Opportunity to clarify definitions and understand how we are using them.
+
+#### Cons
+
+- None!
+
+
+### 3. Replicate the Apply User / Provider Relationship model
```

Comment:
The relationship model in Apply/Manage is complex partly because of safeguarding but also to define who can make application decisions.

Beyond the fundamental relationship (for example, X works with Y), I caution against a complex relationship model that may introduce a permissions model like Manage.

Here is some early work we did on permissions in Publish, which we ended up not using: 
- [Giving lead schools the ability to manage permissions for their accredited bodies](https://becoming-a-teacher.design-history.education.gov.uk/publish-teacher-training-courses/giving-lead-schools-the-ability-to-manage-permissions-for-their-accredited-bodies/)
- [Giving accredited bodies the ability to manage courses they ratify](https://becoming-a-teacher.design-history.education.gov.uk/publish-teacher-training-courses/giving-accredited-bodies-the-ability-to-manage-courses-they-ratify/)

And a related post [Managing accredited bodies if you’re a lead school](https://becoming-a-teacher.design-history.education.gov.uk/publish-teacher-training-courses/managing-accredited-bodies-if-youre-a-lead-school/) which includes changes that were implemented.

(NB. 'accredited body' and 'lead school' are outdated terms replaced by 'accredited provider' and 'lead partner' respectively.)

Reply:
I suppose the Managing accredited bodies was done before the move to "ratifying provider" too.